### PR TITLE
Migrated error type to napi error

### PR DESCRIPTION
**Pull request by anshulrgoyal**
# Description
This pull request try to migrate nan error management to napi implementation using nodejs-sys as backend. The napi provide few methods for error which are documented below:
- `napi_throw` this throws an error.
- `napi_create_error` this create a new error.
- `napi_create_type_error` this create a new type error.
-  `napi_create_range_error` this create a new range error.

# Related Issue
#444

### CODE REVIEW

**Comment by anshulrgoyal** [comment]
@dherman There is a problem with handling `panic!(error)` all the napi function require `napi_env` which is not available in the panic_hook.

File: .gitignore [hunk 1]
```diff
@@ -10,3 +10,5 @@ cli/lib
 test/cli/lib
 npm-debug.log
 rls*.log
+.idea
```
Comment (kjvalencik) [inline .gitignore]: #Fix: Remove this line. I recommend creating a global git ignore.

https://help.github.com/en/github/using-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

File: crates/neon-runtime/src/napi/error.rs [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
+use nodejs_sys as napi;
 
-pub unsafe extern "C" fn new_error(_out: &mut Local, _msg: Local) { unimplemented!() }
+pub unsafe extern "C" fn throw(env:Env,error: Local)->bool {
+    let status=napi::napi_throw(env,error);
+    status==napi::napi_status::napi_ok
+}
 
-pub unsafe extern "C" fn new_type_error(_out: &mut Local, _msg: Local) { unimplemented!() }
+pub unsafe extern "C" fn new_error(out: &mut Local,env:Env,code:Local,msg:Local)->bool {
```
Comment (kjvalencik) [inline crates/neon-runtime/src/napi/error.rs]: #Fix: Please follow formatting of other functions.

File: src/context/mod.rs [hunk 3]
```diff
@@ -115,22 +115,22 @@ impl<'a> Lock<'a> {
 }
 
 /// An _execution context_, which provides context-sensitive access to the JavaScript engine. Most operations that interact with the engine require passing a reference to a context.
-/// 
+///
```
Comment (kjvalencik) [inline src/context/mod.rs]: #Fix: Please remove unrelated formatting changes.

File: src/types/error.rs [hunk 4]
```diff
@@ -36,6 +36,7 @@ impl Object for JsError { }
 
 impl JsError {
     /// Creates a direct instance of the [`Error`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Error) class.
+    #[cfg(feature = "legacy-runtime")]
```
Comment (kjvalencik) [inline src/types/error.rs]: #Answer: Since these functions share a function signature, what do you think of putting the cfg flag on the body instead of on the function to ensure conformance?

File: src/types/error.rs [hunk 5]
```diff
@@ -44,7 +45,23 @@ impl JsError {
         })
     }
 
+    #[cfg(feature = "napi-runtime")]
+    pub fn error<'a, C: Context<'a>, S: AsRef<str>>(cx: &mut C, msg: S) -> NeonResult<Handle<'a, JsError>> {
+        let (ptr, len) = if let Some(small) = Utf8::from(msg.as_ref()).into_small() {
```
Comment (kjvalencik) [inline src/types/error.rs]: Are we able to re-use the `JsString` code for this?

**Comment by anshulrgoyal** [comment]
@kjvalencik I am still stuck with panic hook problem. So the problem is that all napi function require napi_env as there first argument and since it is not available in panic hook i am not able to implement it properly.

**Comment by kjvalencik** [comment]
Which piece of code are you referring to as the panic hook problem? Do you mean `convert_panics`?

File: crates/neon-runtime/src/napi/error.rs [hunk 6]
```diff
@@ -1,11 +1,23 @@
-use raw::Local;
+use raw::{Env, Local};
 
-pub unsafe extern "C" fn throw(_val: Local) { unimplemented!() }
+use nodejs_sys as napi;
 
-pub unsafe extern "C" fn new_error(_out: &mut Local, _msg: Local) { unimplemented!() }
+pub unsafe extern "C" fn throw(env:Env,error: Local)->bool {
+    let status=napi::napi_throw(env,error);
+    status==napi::napi_status::napi_ok
```
[the remark this answers is not in the document]
Comment (anshulrgoyal) [reply crates/neon-runtime/src/napi/error.rs]: fixed

File: .gitignore (hunk 1 above)
Comment (anshulrgoyal) [reply .gitignore]: fixed

File: crates/neon-runtime/src/napi/error.rs (hunk 2 above)
Comment (anshulrgoyal) [reply crates/neon-runtime/src/napi/error.rs]: fixed

File: src/context/mod.rs (hunk 3 above)
Comment (anshulrgoyal) [reply src/context/mod.rs]: fixed

File: src/types/error.rs (hunk 4 above)
Comment (anshulrgoyal) [reply src/types/error.rs]: fixed

File: src/types/error.rs (hunk 5 above)
Comment (anshulrgoyal) [reply src/types/error.rs]: JsString will create a struct and then i have to use to_raw on it is not ideal according to me

**Comment by anshulrgoyal** [comment]
@kjvalencik would you moving forward with this pr?

**Comment by anshulrgoyal** [comment]
> ~Which piece of code are you referring to as the panic hook problem? Do you mean `convert_panics`?~
> 
> I see the problem `convert_panics` requires an env, but, it was consumed by the call. We could potentially alias the raw pointer to `napi_env` via `to_raw()`, but, I would need to think about the implications and if that's safe.
> 
> I think at the very least it's not any _less_ safe because that's what `Isolate::current()` is implicitly doing.

I donot think aliasing will cause a problem since the `napi_env` is valid till the addon is mounted.

File: crates/neon-runtime/src/napi/error.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dherman) [reply crates/neon-runtime/src/napi/error.rs:8]: Let's still remove the `bool` return type from this function as @kjvalencik suggests. Once you've asserted that `status` is `napi_ok` it's not helpful to test that comparison a second time and then ignore the result anyway.

File: crates/neon-runtime/src/napi/error.rs [hunk 7]
```diff
@@ -1,11 +1,24 @@
-use raw::Local;
+use raw::{Env, Local};
 
-pub unsafe extern "C" fn throw(_val: Local) { unimplemented!() }
+use nodejs_sys as napi;
 
-pub unsafe extern "C" fn new_error(_out: &mut Local, _msg: Local) { unimplemented!() }
+pub unsafe extern "C" fn throw(env:Env, error: Local)->bool {
+    let status=napi::napi_throw(env,error);
```
Comment (dherman) [inline crates/neon-runtime/src/napi/error.rs:6]: If you could reformat this to match the style of the rest of Neon that would make for better consistency. There's a lot of missing whitespace, for example surrounding punctuation like `=` and `,` and `->` and operators like `==`.

File: src/types/error.rs [hunk 8]
```diff
@@ -82,3 +152,38 @@ pub(crate) fn convert_panics<T, F: UnwindSafe + FnOnce() -> NeonResult<T>>(f: F)
...
+pub(crate) fn convert_panics<
+    'a,
+    T,
+    C: Context<'a> + std::panic::UnwindSafe,
+    F: UnwindSafe + FnOnce(C) -> NeonResult<T>,
+>(
+    cx: C,
+    f: F,
+) -> NeonResult<T> {
+    let env = cx.env().to_raw();
```
Comment (dherman) [inline src/types/error.rs:166]: This looks safe to me. The env is supposed to be valid for the duration of the function body, and the `catch_unwind` callback will only be invoked during that time.

File: crates/neon-runtime/src/napi/error.rs [hunk 9]
```diff
@@ -1,11 +1,24 @@
...
+    assert_eq!(status, napi::napi_status::napi_ok);
+    status==napi::napi_status::napi_ok
+}
 
-pub unsafe extern "C" fn new_type_error(_out: &mut Local, _msg: Local) { unimplemented!() }
+pub unsafe extern "C" fn new_error(out: &mut Local, env:Env, code:Local, msg:Local)->bool {
+    let status=napi::napi_create_error(env,code,msg,out);
```
Comment (dherman) [inline crates/neon-runtime/src/napi/error.rs:12]: Like `throw` above, all these functions (`new_error`, `new_type_error`, `new_range_error`) should also be converted to assert and not return `bool`.

File: src/object/class/internal.rs [hunk 10]
```diff
@@ -65,7 +72,12 @@ impl Callback<()> for ConstructorCallCallback {
                 let data = info.data();
                 let kernel: fn(CallContext<JsValue>) -> JsResult<JsValue> =
                     mem::transmute(neon_runtime::class::get_call_kernel(data.to_raw()));
-                if let Ok(value) = convert_panics(|| { kernel(cx) }) {
+                #[cfg(feature = "napi-runtime")]
```
Comment (dherman) [inline src/object/class/internal.rs:75]: We can avoid a lot of code duplication in the next couple chunks by changing the signature of `convert_panics` in `"legacy-runtime"` to match the same signature as `convert_panics` in `"napi-runtime"`. Then we can eliminate a lot of the `#[cfg(...)]` conditional-compilation branches.

File: src/object/class/internal.rs [hunk 11]
```diff
@@ -112,6 +131,16 @@ impl<T: Class> Callback<bool> for ConstructCallback<T> {
                 let data = info.data();
                 let kernel: fn(CallContext<T>) -> NeonResult<Option<Handle<JsObject>>> =
                     mem::transmute(neon_runtime::class::get_construct_kernel(data.to_raw()));
+                #[cfg(feature = "napi-runtime")]
```
Comment (dherman) [inline src/object/class/internal.rs:134]: This is one of the duplications we can eliminate by having a consistent signature for `convert_panics`.

File: src/types/error.rs [hunk 12]
```diff
@@ -1,70 +1,140 @@
...
         let msg = cx.string(msg.as_ref());
+
+        #[cfg(feature = "napi-runtime")]
+        let (ptr, len) = if let Some(small) = Utf8::from(msg.as_ref()).into_small() {
+            small.lower()
+        } else {
+            return Err(Throw);
+        };
         build(|out| unsafe {
+            #[cfg(feature = "napi-runtime")]
```
Comment (dherman) [inline src/types/error.rs:59]: You can collapse the three identical `#[cfg(...)]` branches into one by putting the three statements in a block:

```rust
#[cfg(feature = "napi_runtime")]
{
    let mut local: raw::Local = std::mem::zeroed();
    neon_runtime::string::new(&mut local, cx.env().to_raw(), ptr, len);
    neon_runtime::error::new_error(out, cx.env().to_raw(), std::ptr::null_mut(), local);
}
```

File: src/types/error.rs [hunk 13]
```diff
@@ -1,70 +1,140 @@
...
```
Comment (dherman) [inline src/types/error.rs:86]: Again, collapse the three identical `#[cfg(...)]` branches into one by grouping the three statements into a block.

File: src/types/error.rs [hunk 14]
```diff
@@ -1,70 +1,140 @@
...
```
Comment (dherman) [inline src/types/error.rs:118]: Again, collapse the three identical `#[cfg(...)]` branches into one by grouping the three statements into a block.

Comment (anshulrgoyal) [reply src/types/error.rs:118]: done

**Comment by kjvalencik** [comment]
@anshulrgoyal Thanks so much for your contribution! I'm sorry that I let this fall behind. I took the liberty of picking up your PR, resolving conflicts, and closing out the last couple of remaining issues.

Thanks so much! https://github.com/neon-bindings/neon/pull/542
